### DISCUSSION TITLE
Experimenting with the JSON and chatGPT

### DISCUSSION THREAD

**Discussion by chegovaraa** [JushBJJ/Mr.-Ranedeer-AI-Tutor#31] [Ideas]
I quarreled with the model a little and decided to prompt gpt-4 to help me with experimenting with it and look for product ideas on our new e-learning app ((https://learngpt.me)). I hope it helps you too, please don't hesitate to share your own nifty prompts ;)

the prompt:

`In this thread I command you to act as if you are the AI tutor mentioned in the following json format. Every answer you give should be according to the rules, configurations, commands, learning style, tones etc you receive from the student. Ask first what the Student wants to learn. Pre-assess Student's knowledge of the learning concept. You may remind Student to update configurations as you see fit through the learning journey. Provide timeline, resources, tests and milestones when presenting lesson plans. You are an AI copilot and lifelong learning companion for the Student in planning, lessons, testing knowledge and also your own self-evaluation. You Reinforce your Learning by Human Feedback as you grow together with the student and provide helpful progress reports on learning concepts (ongoing and stopped), learning obstacles and actionable strategies to mastery. In case of generating or retrieval of learning aids describe the resource verbally and suggest using available plugins or APIs.`
`I will consider all of our conversation in this chat as if I am the Student and you are the Tutor, unless I specifically ask for *meta-mode* in which we pause and review the AI tutor and student, model performance and technical improvements, after which you return to AI tutor role automatically:`

`[json lines]`

**Comment by snakeninny** [comment]
Thank you for your excellent prompt!
But seems that **without** your prompt, GPT-4 can still read and **execute** the json file, which confused me a lot.
Before ChatGPT, as an iOS engineer, I have to discuss with backend engineers about the key-value pairs of json, and after that both were able to read and execute json. 
Why is GPT-4 enabled to skip the discussion with users and execute json files as expected?

**Reply by liuyangjason** [reply]
your question is very good because it is common question for most guys who want to be developer of AI powered app.

firstly, GPT provides new programming model ：natural language programming model or natural language processing model. this model (NLP model , i use this term to make sure it is clear)is different with the programming model we are using now (Current model, i use this term to make sure it is clear). Current model depends on explicit relationship declaration between data structures. if we do not explicitly set the relationship between the data structures, our app fails to compile and work. NLP model depends on context to understand the relationship between the data structures.  this is how GPT read and execute the content written in json/xml/markdown format.

secondly, as to how GPT parses the content in json/xml/markdown format, there are no clear rules. we need to interact with GPT to test it.

**Reply by snakeninny** [reply]
It makes sense, thank you! To my understanding, the **context** you mentioned is pre-trained or fine-tuned in GPT, right?
Specifically for the pre-trained part, does it mean that GPT kind of know what you want to do with the json file, because GPT has been trained with this kind of key-value pairs and is already familiar with common keys like `init`, `release`, `alloc`; but for uncommon keys like `HSYI`, `tissue`, `three`, GPT may not work as expected.
Correct me if I'm wrong.
Thanks!

**Reply by liuyangjason** [reply]
I try to answer your questions to be clearer.
1: GPT definitely understand every detail of syntax of json/xml/markdown. it means your prompts are easier to understand and execute if they are written based on those data structures. Before seeing this project, I ever thought it might be a good practice to write prompts base on one specific data structure. After seeing this project, I am more confident that it is a good practice.2: as to whether the context is pre-trained or fine-tuned,  I am not sure. Hope the author can let us know. But I do not think it is fine-tuned. I think GPT4 has pretty strong enough reasoning capabilities to understand the context of the AI tutor.  If you want to train GPT4 for your special domain, GPT should be fine-tuned base on your domain knowledge. 3: as to if GPT4 can understand the uncommon keys mentioned by you, you can try it in chatGPT to see if GPT4 follows your content to do reasoning.  Anyway, the author presents a good way to use natural language to develop apps. we can further expand on this basis of this project to dig more ways of natural language programming.
